File: ui-bookmarks/src/main/java/l/files/ui/bookmarks/BookmarksFragment.java
package l.files.ui.bookmarks;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.view.ActionMode;
import androidx.lifecycle.Observer;
import androidx.lifecycle.Transformations;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import l.files.bookmarks.BookmarksManager;
import l.files.bookmarks.BookmarksKt;
import l.files.fs.Path;
import l.files.ui.base.fs.OpenFileEvent;
import l.files.ui.base.selection.SelectionModeFragment;
import l.files.ui.base.view.ActionModeProvider;
import l.files.ui.base.view.ActionModes;
import l.files.ui.base.view.ClearSelectionOnDestroyActionMode;
import l.files.ui.base.view.CountSelectedItemsAction;

import java.text.Collator;
import java.util.ArrayList;
import java.util.List;

import static kotlin.collections.CollectionsKt.filterIsInstance;
import static l.files.bookmarks.BookmarksKt.getBookmarkManager;
import static l.files.ui.base.fs.UserDirs.DIR_HOME;

public final class BookmarksFragment
        extends SelectionModeFragment<Path, Path>
        implements Observer<List<Path>> {

    public RecyclerView recycler;
    private BookmarksAdapter adapter;
    private BookmarksManager bookmarks;

    public List<Path> bookmarks() {
        return filterIsInstance(adapter.items(), Path.class);
    }

    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState
    ) {
        return inflater.inflate(R.layout.bookmarks_fragment, container, false);
    }

    @Override
    public void onViewCreated(
            @NonNull View view,
            @Nullable Bundle savedInstanceState
    ) {
        super.onViewCreated(view, savedInstanceState);

        bookmarks = getBookmarkManager(this);

        Transformations.map(
                bookmarks.getLiveData(),
                paths -> BookmarksKt.collate(
                        paths,
                        path -> path.equals(DIR_HOME),
                        Collator.getInstance()
                )
        ).observe(getViewLifecycleOwner(), this);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        adapter = new BookmarksAdapter(
                selection(),
                actionModeProvider(),
                actionModeCallback(),
                OpenFileEvent.topic
        );

        View view = getView();
        assert view != null;

        recycler = view.findViewById(R.id.bookmarks);
        recycler.setLayoutManager(new LinearLayoutManager(getActivity()));
        recycler.setAdapter(adapter);
    }

    @Override
    protected ActionMode.Callback actionModeCallback() {
        return ActionModes.compose(
                new CountSelectedItemsAction(selection()),
                new ClearSelectionOnDestroyActionMode(selection()),
                new RemoveBookmark(selection(), bookmarks)
        );
    }

    @Override
    protected ActionModeProvider actionModeProvider() {
        return (ActionModeProvider) getActivity();
    }

    @Override
    public void onChanged(List<Path> bookmarks) {
        List<Object> items = new ArrayList<>(bookmarks.size() + 1);
        items.add(getString(R.string.bookmarks));
        items.addAll(bookmarks);
        adapter.setItems(items);
    }
}

File: ui-bookmarks/src/main/java/l/files/ui/bookmarks/BookmarksFragment.kt
package l.files.ui.bookmarks

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.view.ActionMode
import androidx.lifecycle.Observer
import androidx.lifecycle.map
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.bookmarks_fragment.view.*
import l.files.base.lifecycle.SetLiveData
import l.files.bookmarks.collate
import l.files.bookmarks.getBookmarks
import l.files.ui.base.fs.OpenFileEvent
import l.files.ui.base.fs.UserDirs.DIR_HOME
import l.files.ui.base.selection.SelectionModeFragment
import l.files.ui.base.view.ActionModeProvider
import l.files.ui.base.view.ActionModes
import l.files.ui.base.view.ClearSelectionOnDestroyActionMode
import l.files.ui.base.view.CountSelectedItemsAction
import java.nio.file.Path

class BookmarksFragment :
  SelectionModeFragment<Path, Path>(),
  Observer<List<Path>> {

  lateinit var recycler: RecyclerView
  private lateinit var adapter: BookmarksAdapter
  private lateinit var bookmarks: SetLiveData<Path>

  fun bookmarks(): List<Path> = adapter.items()
    .filterIsInstance<Path>()

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View = inflater.inflate(R.layout.bookmarks_fragment, container, false)

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    bookmarks = getBookmarks()
    bookmarks
      .map { it.collate({ path -> path == DIR_HOME }) }
      .observe(viewLifecycleOwner, this)
  }

  override fun onActivityCreated(savedInstanceState: Bundle?) {
    super.onActivityCreated(savedInstanceState)
    adapter = BookmarksAdapter(
      selection(),
      actionModeProvider(),
      actionModeCallback(),
      OpenFileEvent.topic
    )
    recycler = requireView().bookmarks
    recycler.layoutManager = LinearLayoutManager(activity)
    recycler.adapter = adapter
  }

  override fun actionModeCallback(): ActionMode.Callback =
    ActionModes.compose(
      CountSelectedItemsAction(selection()),
      ClearSelectionOnDestroyActionMode(selection()),
      RemoveBookmark(selection(), bookmarks)
    )

  override fun actionModeProvider() = activity as ActionModeProvider

  override fun onChanged(bookmarks: List<Path>) {
    adapter.setItems(listOf(getString(R.string.bookmarks)) + bookmarks)
  }
}
